README.md
String Score [![Build Status](https://travis-ci.org/bcse/stringscore.png?branch=master)](https://travis-ci.org/bcse/stringscore)
============

An algorithm provides scores between 0.0 (no match) to 1.0 (perfect match) for a comparison of two strings.

The algorithm is designed for auto-completion. For string similarity, please check [Levenshtein distance](https://en.wikipedia.org/wiki/Levenshtein_distance) ([Wagner–Fischer algorithm](https://en.wikipedia.org/wiki/Wagner-Fischer_algorithm)).

Usage
-----

Include the library:

```python
from stringscore import liquidmetal
```

Score any string against an abbreviation:

```python
>>> liquidmetal.score('FooBar', 'foo')
0.95
>>> liquidmetal.score('FooBar', 'fb')
0.916666666667
>>> liquidmetal.score('Foo Bar', 'fb')
0.928571428571
>>> liquidmetal.score('Foo Bar', 'baz')
0.0
>>> liquidmetal.score('Foo Bar', '')
0.8
```

Similar Works
-------------

- Quicksilver's [scoreForAbbreviation](https://github.com/quicksilver/Quicksilver/blob/master/Quicksilver/Code-QuickStepFoundation/NSString_BLTRExtensions.m#L53) algorithm by Alcor (Blacktree, Inc)
- [LiquidMetal](https://github.com/rmm5t/liquidmetal) by Ryan McGeary
- [string_score](https://github.com/joshaven/string_score) by Joshaven Potter
- [jQuery.fuzzyMatch](https://github.com/rapportive-oss/jquery-fuzzymatch) by [Rapportive](http://rapportive.com/)

License
-------

String Score is released under the [MIT License](http://opensource.org/licenses/MIT).

Credits
-------

Copyright © 2003 Blacktree, Inc (Original author of [Quicksilver](https://github.com/quicksilver/Quicksilver))  
Copyright © 2009 Ryan McGeary (Author of [LiquidMetal](https://github.com/rmm5t/liquidmetal))  
Copyright © 2013 Grey Lee


[![Bitdeli Badge](https://d2weczhvl823v0.cloudfront.net/bcse/stringscore/trend.png)](https://bitdeli.com/free "Bitdeli Badge")



HISTORY.rst
.. :changelog:

History
-------

0.1.0 (2013-02-25)
++++++++++++++++++

* First release on PyPI.

test_stringscore.py
import unittest

from stringscore import liquidmetal
from stringscore import quicksilver


class LiquidMetalTestCase(unittest.TestCase):

    def test_score(self):
        n = liquidmetal.SCORE_NO_MATCH
        m = liquidmetal.SCORE_MATCH
        t = liquidmetal.SCORE_TRAILING
        s = liquidmetal.SCORE_TRAILING_BUT_STARTED
        b = liquidmetal.SCORE_BUFFER

        tests = {
            ('', ''): [t],
            ('', 'a'): [n],
            ('a', ''): [t],
            ('a', 'toolong'): [n],
            ('a', 'a'): [m],
            ('a', 'b'): [n],
            ('abc', ''): [t, t, t],
            ('abc', 'a'): [m, s, s],
            ('abc', 'b'): [n, m, t],
            ('abc', 'c'): [n, n, m],
            ('abc', 'd'): [n, n, n],
            ('A', 'a'): [m],
            ('A', 'b'): [n],
            ('FooBar', ''): [t, t, t, t, t, t],
            ('FooBar', 'foo'): [m, m, m, s, s, s],
            ('FooBar', 'fb'): [m, b, b, m, s, s],
            ('foobar', 'fb'): [m, n, n, m, s, s],
            ('FooBar', 'b'): [b, b, b, m, t, t],
            ('FooBar', 'ooar'): [n, m, m, n, m, m],
            ('FooBar', 'bab'): [n, n, n, n, n, n],
            ('Foo Bar', ''): [t, t, t, t, t, t, t],
            ('Foo Bar', 'foo'): [m, m, m, s, s, s, s],
            ('Foo Bar', 'fb'): [m, b, b, m, m, s, s],
            ('Foo-Bar', 'fb'): [m, b, b, m, m, s, s],
            ('Foo_Bar', 'fb'): [m, b, b, m, m, s, s],
            ('Foo Bar', 'b'): [b, b, b, m, m, t, t],
            ('Foo Bar', 'ooar'): [n, m, m, n, n, m, m],
            ('Foo Bar', 'bab'): [n, n, n, n, n, n, n],
            ('gnu\'s Not Unix', 'nu'): [b, b, b, b, b, m, m,
                                        b, b, m, m, t, t, t],
        }

        for k, v in tests.items():
            score = round(liquidmetal.score(*k), 12)
            expected_score = round(sum(v) / len(v), 12)
            self.assertEqual(score, expected_score)


class QuickSilverTestCase(unittest.TestCase):

    def test_score(self):
        tests = [
            (('', ''), 0.9),
            (('', 'a'), 0.0),
            (('a', ''), 0.9),
            (('a', 'toolong'), 0.0),
            (('a', 'a'), 1.0),
            (('a', 'b'), 0.0),
            (('abc', ''), 0.9),
            (('abc', 'a'), 0.933333333333),
            (('abc', 'b'), 0.633333333333),
            (('abc', 'c'), 0.333333333333),
            (('abc', 'd'), 0.0),
            (('A', 'a'), 1.0),
            (('A', 'b'), 0.0),
            (('FooBar', ''), 0.9),
            (('FooBar', 'foo'), 0.95),
            (('FooBar', 'fb'), 0.916666666667),
            (('foobar', 'fb'), 0.633333333333),
            (('FooBar', 'b'), 0.75),
            (('FooBar', 'ooar'), 0.666666666667),
            (('FooBar', 'bab'), 0.0),
            (('Foo Bar', ''), 0.9),
            (('Foo Bar', 'foo'), 0.942857142857),
            (('Foo Bar', 'fb'), 0.928571428571),
            (('Foo-Bar', 'fb'), 0.907142857143),
            (('Foo_Bar', 'fb'), 0.907142857143),
            (('Foo Bar', 'b'), 0.907142857143),
            (('Foo Bar', 'ooar'), 0.571428571429),
            (('Foo Bar', 'bab'), 0.0),
            (('gnu\'s Not Unix', 'nu'), 0.85),
        ]

        for k, v in tests:
            self.assertEqual(round(quicksilver.score(*k), 12), v)


if __name__ == '__main__':
    unittest.main()


tests/test_stringscore.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
test_stringscore
----------------------------------

Tests for `stringscore` module.
"""

import unittest

from stringscore import liquidmetal
from stringscore import quicksilver


class LiquidMetalTestCase(unittest.TestCase):

    def test_score(self):
        n = liquidmetal.SCORE_NO_MATCH
        m = liquidmetal.SCORE_MATCH
        t = liquidmetal.SCORE_TRAILING
        s = liquidmetal.SCORE_TRAILING_BUT_STARTED
        b = liquidmetal.SCORE_BUFFER

        tests = {
            ('', ''): [t],
            ('', 'a'): [n],
            ('a', ''): [t],
            ('a', 'toolong'): [n],
            ('a', 'a'): [m],
            ('a', 'b'): [n],
            ('abc', ''): [t, t, t],
            ('abc', 'a'): [m, s, s],
            ('abc', 'b'): [n, m, t],
            ('abc', 'c'): [n, n, m],
            ('abc', 'd'): [n, n, n],
            ('A', 'a'): [m],
            ('A', 'b'): [n],
            ('FooBar', ''): [t, t, t, t, t, t],
            ('FooBar', 'foo'): [m, m, m, s, s, s],
            ('FooBar', 'fb'): [m, b, b, m, s, s],
            ('foobar', 'fb'): [m, n, n, m, s, s],
            ('FooBar', 'b'): [b, b, b, m, t, t],
            ('FooBar', 'ooar'): [n, m, m, n, m, m],
            ('FooBar', 'bab'): [n, n, n, n, n, n],
            ('Foo Bar', ''): [t, t, t, t, t, t, t],
            ('Foo Bar', 'foo'): [m, m, m, s, s, s, s],
            ('Foo Bar', 'fb'): [m, b, b, m, m, s, s],
            ('Foo-Bar', 'fb'): [m, b, b, m, m, s, s],
            ('Foo_Bar', 'fb'): [m, b, b, m, m, s, s],
            ('Foo Bar', 'b'): [b, b, b, m, m, t, t],
            ('Foo Bar', 'ooar'): [n, m, m, n, n, m, m],
            ('Foo Bar', 'bab'): [n, n, n, n, n, n, n],
            ('gnu\'s Not Unix', 'nu'): [b, b, b, b, b, m, m,
                                        b, b, m, m, t, t, t],
        }

        for k, v in tests.items():
            score = round(liquidmetal.score(*k), 12)
            expected_score = round(sum(v) / len(v), 12)
            self.assertEqual(score, expected_score)


class QuickSilverTestCase(unittest.TestCase):

    def test_score(self):
        tests = [
            (('', ''), 0.9),
            (('', 'a'), 0.0),
            (('a', ''), 0.9),
            (('a', 'toolong'), 0.0),
            (('a', 'a'), 1.0),
            (('a', 'b'), 0.0),
            (('abc', ''), 0.9),
            (('abc', 'a'), 0.933333333333),
            (('abc', 'b'), 0.633333333333),
            (('abc', 'c'), 0.333333333333),
            (('abc', 'd'), 0.0),
            (('A', 'a'), 1.0),
            (('A', 'b'), 0.0),
            (('FooBar', ''), 0.9),
            (('FooBar', 'foo'), 0.95),
            (('FooBar', 'fb'), 0.916666666667),
            (('foobar', 'fb'), 0.633333333333),
            (('FooBar', 'b'), 0.75),
            (('FooBar', 'ooar'), 0.666666666667),
            (('FooBar', 'bab'), 0.0),
            (('Foo Bar', ''), 0.9),
            (('Foo Bar', 'foo'), 0.942857142857),
            (('Foo Bar', 'fb'), 0.928571428571),
            (('Foo-Bar', 'fb'), 0.907142857143),
            (('Foo_Bar', 'fb'), 0.907142857143),
            (('Foo Bar', 'b'), 0.907142857143),
            (('Foo Bar', 'ooar'), 0.571428571429),
            (('Foo Bar', 'bab'), 0.0),
            (('gnu\'s Not Unix', 'nu'), 0.85),
        ]

        for k, v in tests:
            self.assertEqual(round(quicksilver.score(*k), 12), v)


if __name__ == '__main__':
    unittest.main()


stringscore/liquidmetal.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# Ported from LiquidMetal
# https://github.com/rmm5t/liquidmetal

SCORE_NO_MATCH = 0.0
SCORE_MATCH = 1.0
SCORE_TRAILING = 0.8
SCORE_TRAILING_BUT_STARTED = 0.9
SCORE_BUFFER = 0.85
WORD_SEPARATORS = ' \t_-'


def score(string, abbrev):
    # short circuits
    if len(string) < len(abbrev):
        # string, abbrev = abbrev, string
        return SCORE_NO_MATCH

    if len(abbrev) == 0:
        return SCORE_TRAILING

    # match & score all
    all_scores = []
    _score_all(string, string.lower(), abbrev.lower(),
               -1, 0, [None] * len(string), all_scores)

    # complete miss
    if len(all_scores) == 0:
        return SCORE_NO_MATCH

    # sum per-character scores into overall scores,
    # selecting the maximum score
    max_score = 0.0
    for i in range(len(all_scores)):
        scores = all_scores[i]
        score_sum = sum(scores)
        if score_sum > max_score:
            max_score = score_sum

    # normalize max score by string length
    # s. t. the perfect match score = 1
    max_score /= len(string)

    # record maximum score & score array, return
    return max_score


def _score_all(string, search, abbrev, search_index, abbr_index, scores,
               all_scores):
    # save completed match scores at end of search
    if abbr_index == len(abbrev):
        # add trailing score for the remainder of the match
        started = search[0] == abbrev[0]
        trail_score = SCORE_TRAILING_BUT_STARTED if started else SCORE_TRAILING
        if None in scores:
            begin = scores.index(None)
            end = len(scores) - scores[::-1].index(None)
            fill_array(scores, trail_score, begin, end)
        # save score clone (since reference is persisted in scores)
        all_scores.append(scores[:])
        return

    # consume current char to match
    c = abbrev[abbr_index]
    abbr_index += 1

    # cancel match if a character is missing
    try:
        index = search.index(c, search_index + 1)
    except ValueError:
        return

    # match all instances of the abbreviaton char
    score_index = search_index  # score section to update
    while True:
        try:
            index = search.index(c, search_index + 1)
        except ValueError:
            break

        # score this match according to context
        if is_new_word(string, index):
            if index > 0:
                scores[index - 1] = SCORE_MATCH
                fill_array(scores, SCORE_BUFFER, score_index + 1, index - 1)
        elif is_upper_case(string, index):
            fill_array(scores, SCORE_BUFFER, score_index + 1, index)
        else:
            fill_array(scores, SCORE_NO_MATCH, score_index + 1, index)

        scores[index] = SCORE_MATCH

        # consume matched string and continue search
        search_index = index
        _score_all(string, search, abbrev, search_index,
                   abbr_index, scores, all_scores)


def is_upper_case(string, index):
    c = string[index]
    return 'A' <= c <= 'Z'


def is_new_word(string, index):
    c = string[index - 1] if index > 0 else ''
    return c in WORD_SEPARATORS


def fill_array(array, value, begin, end):
    if end > begin:
        array[begin:end] = [value] * (end - begin)


if __name__ == '__main__':
    from time import clock
    t0 = clock()
    test_string = 'a'
    for string in open('tests/words'):
        abbrev = test_string
        test_string = string
        if len(abbrev) > len(string):
            string, abbrev = abbrev, string
        score(string, abbrev)
    print('Benchmark: %ss' % (clock() - t0))


stringscore/quicksilver.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# Ported from Quicksilver's scoreForAbbreviation algorithm
# https://github.com/quicksilver/Quicksilver/

SCORE_NO_MATCH = 0.0
SCORE_MATCH = 1.0
SCORE_TRAILING = 0.9
SCORE_BUFFER = 0.15
WHITESPACE_CHARACTERS = ' \t'


def score(string, abbrev):
    abbrev = abbrev.lower()
    abbrev_len = len(abbrev)
    string_len = len(string)

    # deduct some points for all remaining letters
    if abbrev_len == 0:
        return SCORE_TRAILING
    if abbrev_len > string_len:
        return SCORE_NO_MATCH

    # Search for steadily smaller portions of the abbreviation
    for i in range(abbrev_len, 0, -1):
        try:
            index = string.lower().index(abbrev[:i])
        except ValueError:
            continue  # Not found

        if index + abbrev_len > string_len:
            continue

        next_string = string[index + i:]
        next_abbrev = abbrev[i:]

        # Search what is left of the string with the rest of the abbreviation
        remaining_score = score(next_string, next_abbrev)

        if remaining_score > 0:
            result_score = index + i

            # ignore skipped characters if is first letter of a word
            if index > 0:  # if some letters were skipped
                if string[index - 1] in WHITESPACE_CHARACTERS:
                    for j in range(index - 1):
                        c = string[j]
                        result_score -= \
                            SCORE_MATCH \
                            if c in WHITESPACE_CHARACTERS \
                            else SCORE_BUFFER
                elif 'A' <= string[index] <= 'Z':
                    for j in range(index):
                        c = string[j]
                        result_score -= \
                            SCORE_MATCH \
                            if 'A' <= c <= 'Z' \
                            else SCORE_BUFFER
                else:
                    result_score -= index

            result_score += remaining_score * len(next_string)
            result_score /= string_len
            return result_score

    return SCORE_NO_MATCH


if __name__ == '__main__':
    from time import clock
    t0 = clock()
    test_string = 'a'
    for string in open('tests/words'):
        abbrev = test_string
        test_string = string
        if len(abbrev) > len(string):
            string, abbrev = abbrev, string
        score(string, abbrev)
    print('Benchmark: %ss' % (clock() - t0))


stringscore.py
"""
String Score
~~~~~~~~~~~~

Port of [LiquidMetal](https://github.com/rmm5t/liquidmetal) from JavaScript to Python.

An algorithm provides scores between 0.0 (no match) to 1.0 (perfect match) for a comparison of two strings.

:copyright: (c) 2013 by Grey Lee.
:license: MIT License.
"""

__version__ = '1.2.1'
__author__ = 'Grey Lee <bcse@bcse.tw>'


SCORE_NO_MATCH = 0.0
SCORE_MATCH = 1.0
SCORE_TRAILING = 0.8
SCORE_TRAILING_BUT_STARTED = 0.9
SCORE_BUFFER = 0.85
WORD_SEPARATORS = ' \t_-'


def score(string, abbrev):
    # short circuits
    if len(string) < len(abbrev):
        # string, abbrev = abbrev, string
        return SCORE_NO_MATCH

    if len(abbrev) == 0:
        return SCORE_TRAILING

    # match & score all
    all_scores = []
    _score_all(string, string.lower(), abbrev.lower(), -1, 0, [None] * len(string), all_scores)

    # complete miss
    if len(all_scores) == 0:
        return SCORE_NO_MATCH

    # sum per-character scores into overall scores,
    # selecting the maximum score
    max_score = 0.0
    for i in range(len(all_scores)):
        scores = all_scores[i]
        score_sum = sum(scores)
        if score_sum > max_score:
            max_score = score_sum

    # normalize max score by string length
    # s. t. the perfect match score = 1
    max_score /= len(string)

    # record maximum score & score array, return
    return max_score


def _score_all(string, search, abbrev, search_index, abbr_index, scores, all_scores):
    # save completed match scores at end of search
    if abbr_index == len(abbrev):
        # add trailing score for the remainder of the match
        started = search[0] == abbrev[0]
        trail_score = SCORE_TRAILING_BUT_STARTED if started else SCORE_TRAILING
        if None in scores:
            begin = scores.index(None)
            end = len(scores) - scores[::-1].index(None)
            fill_array(scores, trail_score, begin, end)
        # save score clone (since reference is persisted in scores)
        all_scores.append(scores[:])
        return

    # consume current char to match
    c = abbrev[abbr_index]
    abbr_index += 1

    # cancel match if a character is missing
    try:
        index = search.index(c, search_index + 1)
    except ValueError:
        return

    # match all instances of the abbreviaton char
    score_index = search_index  # score section to update
    while True:
        try:
            index = search.index(c, search_index + 1)
        except ValueError:
            break

        # score this match according to context
        if is_new_word(string, index):
            if index > 0:
                scores[index - 1] = SCORE_MATCH
                fill_array(scores, SCORE_BUFFER, score_index + 1, index - 1)
        elif is_upper_case(string, index):
            fill_array(scores, SCORE_BUFFER, score_index + 1, index)
        else:
            fill_array(scores, SCORE_NO_MATCH, score_index + 1, index)

        scores[index] = SCORE_MATCH

        # consume matched string and continue search
        search_index = index
        _score_all(string, search, abbrev, search_index, abbr_index, scores, all_scores)


def is_upper_case(string, index):
    c = string[index]
    return 'A' <= c <= 'Z'


def is_new_word(string, index):
    c = string[index - 1] if index > 0 else ''
    return c in WORD_SEPARATORS


def fill_array(array, value, begin, end):
    if end > begin:
        array[begin:end] = [value] * (end - begin)


if __name__ == '__main__':
    from time import clock
    t0 = clock()
    test_string = 'a'
    for string in open('/usr/share/dict/words'):
        abbrev = test_string
        test_string = string
        if len(abbrev) > len(string):
            string, abbrev = abbrev, string
        score(string, abbrev)
    print 'Benchmark: %ss' % (clock() - t0)
